### ISSUE TITLE
set_default_device/torch.device has performance impact for non-factory functions

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#92701]
### 🐛 Describe the bug

Because it is implemented as a TorchFunctionMode, we interpose ALL functions even if we actually do nothing on those functions. We have ideas for how to optimize this but have delayed doing so until someone actually has a problem.  This is a placeholder issue to track people having this problem.

### Versions

master

**Comment by dilililiwhy** [comment]
After setting default device by TorchFunctionMode, performance decrease is discovered in training of BiLSTM which contains a large amount of small ops calling no need for override (e.g. 10000+ calling of aten::select).

Is it availiable to interpose specific func?

**Comment by ezyang** [comment]
One thing to try is you often only need the context manager on model loading

**Comment by ani300** [comment]
In our code, we observed a degradation of an aten::mm call from 0.036ms between call and kernel launch to about 0.162ms, so about 4.5x times slower. For an NLP decoding task with small batch sizes, this adds significant latency.

**Comment by youkaichao** [comment]
didn't know `set_default_device` in pytorch has such a huge cost. What if I directly call `cudaSetDevice` from cuda runtime API? will it affect pytorch? @ezyang

**Comment by ezyang** [comment]
cudaSetDevice does something different haha, it changes the default cuda device, which will do something if you're saying device="cuda"

**Comment by youkaichao** [comment]
kind of confused here. cuda has the concept of current device, and I think there are three cases:

1. user specify the device, e.g. `torch.ones(5, device="cuda:1")`, then pytorch should switch the current device to 1, create the tensor, and switch back to the last current device
2. user specify the device without index, e.g. `torch.ones(5, device="cuda")`, then pytorch should just use the current device
3. user does not specify the device, e.g. `torch.ones(5)`, then pytorch should just use cpu

Then I think `with device:` does something similar to 1. But actually I find it is not:

```python
import torch
device = torch.device("cuda:1")
with device:
    data = torch.ones(5).cuda()
    print(data.device) # cuda:0
    print(torch.cuda.current_device()) # 0
```

**Comment by ezyang** [comment]
Yes, with device does something different, which is "whenever there is a PyTorch factory function call, pretend as if you had passed in this device as a device= argument to it"

**Comment by youkaichao** [comment]
seems like we also need a function (context manager) to switch the current device:

```python
@contextmanager
def switch_device(device):
        current_device = torch.cuda.current_device()
        try:
            torch.cuda.set_device(device)
            yield
        finally:
            torch.cuda.set_device(current_device)
```

Any existing one in pytorch that does this? I feel like we should have one, because when we call `torch.ones(5, device="cuda:1")` , it will switch the context to cuda 1, and switch back after the function call.

**Comment by ezyang** [comment]
That's what torch.cuda.device does ;)

**Comment by youkaichao** [comment]
Uhh thanks, didn't know the difference between `with torch.device` and `with torch.cuda.device`. Now I know it.
